### ISSUE TITLE
Mix advertised & attr FEC mode?

### ISSUE THREAD

**Issue by shihjeff** [opencomputeproject/SAI#633]
Hi All,

There has been two different interpretations of the use of below port attributes and we wanted to query the group to see what is the "expected" behavior:
SAI_PORT_ATTR_ADVERTISED_FEC_MODE
SAI_PORT_ATTR_FEC_MODE

Interpretation #1:
They are mutually exclusive and cannot be combined in the same call.  SAI_PORT_ATTR_ADVERTISED_FEC_MODE is for AN and SAI_PORT_ATTR_FEC_MODE is for fixed mode.  Combining both doesn't make sense since fixed mode & AN are two separate modes.  Especially in the 25G/50G scenario where you could have multiple FEC modes:

SAI_PORT_ATTR_ADVERTISED_FEC_MODE == SAI_PORT_FEC_MODE_RS && SAI_PORT_FEC_MODE_FC
SAI_PORT_ATTR_FEC_MODE == SAI_PORT_FEC_MODE_RS

What would above example result in?  

Interpretation #2: 
Always set both of them at the same time for AN.  SAI_PORT_ATTR_FEC_MODE is the priority for FEC AN since there is no obvious "priority" between FC & RS-FEC.  So in above example, we would AN RS/FC but RS-FEC wins if both RS & FC get negotiated.  
If only SAI_PORT_ATTR_FEC_MODE is set alone, it is fixed mode.

Long story short, if one passes multiple attributes, does SAI expect to process them one by one in order; or does it have to do a summation of them (via some pre-processing) and do non-conflicting actions that is not exactly any of attributes individually.  

So in interpretation #2 
1) SAI_PORT_ATTR_FEC_MODE by itself would mean to hardcode the FEC.  
2) SAI_PORT_ATTR_FEC_MODE && SAI_PORT_ATTR_ADVERTISED_FEC_MODE, SAI_PORT_ATTR_FEC_MODE means it is FEC priority for SAI_PORT_ATTR_ADVERTISED_FEC_MODE.

**Comment by shihjeff** [comment]
@lguohan - can you comment?  Especially the model of processing multiple attributes (individually or summation)

Typically most folks assume RS is higher priority than FC if both are advertised.  Is this a safe assumption since RS is newer than FC?

In interpretation #2, instead of re-using an existing enum for the purpose of FEC prioritization (If user wants to set FC > RS), would it be better to create a new enum?

**Comment by shihjeff** [comment]
I think most folks are more into interpretation #1.

This means that for folks who want control of the "request" bits for AN, there is no attribute.  Should we create a new one like SAI_PORT_ATTR_ADVERTISED_FEC_REQUEST_MODE and maybe change existing to SAI_PORT_ATTR_ADVERTISED_FEC_ABILITY_MODE to make it explicit what part of AN process this is controlling

since SAI_PORT_ATTR_ADVERTISED_FEC_MODE == AN ability bits currently

**Comment by shihjeff** [comment]
I've created PR#640 to address this issue.
https://github.com/opencomputeproject/SAI/pull/640

I've added the request bit as an explicit parameter to be toggled and added comment about ability bit implicitly set if FEC is advertised.
